<a href="https://colab.research.google.com/github/ShilpaVasista/Exploratory-Data-Analytics/blob/main/EDA_Lab_9_Case_Study.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Case Study
Movie Ratings and Reviews: An EDA Case Study

Dataset Description:

Imagine you have a dataset containing the following information about movies and user reviews:

- movie_id: Unique identifier for each movie.
- title: Movie title.
- genre: Genre of the movie (e.g., Comedy, Drama, Action).
- director: Director of the movie (categorical).
- release_year: Year the movie was released (numerical).
- user_id: Unique identifier for each user who submitted a review (categorical).
- rating: User rating for the movie (numerical scale).


## Questions for Laboratory Practice



### **Module 1: Introduction to EDA**

**1.** Identify the measurement scales (Nominal, Ordinal, Interval, or Ratio) for the following columns: `movie_id`, `rating`, `age`, and `genre`. Explain your reasoning. **(RBT Level 1: Remember)**

**2.** Load the `movie_ratings_dirty.csv` using Pandas. Write a script to identify the number of missing (`NaN`) values in each column and calculate the percentage of data missing for the `rating` column. **(RBT Level 2: Understand)**

**3.** Apply descriptive statistics (`describe()`) to the numerical features. Based on the output, identify at least two data anomalies (e.g., impossible ages or ratings) that require cleaning. **(RBT Level 3: Apply)**

**4.** Explain the significance of the "Data Refactoring" step in this case study. Why should columns like `movie_id` and `user_id` be handled differently than `rating` during statistical analysis? **(RBT Level 2: Understand)**

**5.** Demonstrate how to drop columns that are not useful for a specific analysis (e.g., dropping `director` for a user-centric demographic study) using the `drop()` method. **(RBT Level 3: Apply)**

---

### **Module 2: Data Transformation**

**6.** The dataset contains duplicate entries. Use the `duplicated()` and `drop_duplicates()` methods to perform data deduplication and verify the count before and after. **(RBT Level 3: Apply)**

**7.** You have a separate `Users` table and a `Movies` table. Demonstrate the use of an **Inner Join** versus a **Left Join** to combine user demographics with their specific movie ratings. **(RBT Level 4: Analyze)**

**8.** Handle missing data in the `rating` column using three different techniques: (a) Dropping rows, (b) Mean Imputation, and (c) Backward/Forward filling. Discuss which method is most appropriate here. **(RBT Level 4: Analyze)**

**9.** Perform data standardization on the `genre` and `gender` columns. Use the `replace()` or `map()` method to fix inconsistencies (e.g., converting "sci-fi" to "Sci-Fi" and "M" to "Male"). **(RBT Level 3: Apply)**

**10.** **Discretization:** Create a new column `age_group` by binning the `age` column into four categories: 'Teen', 'Young Adult', 'Middle Aged', and 'Senior'. **(RBT Level 3: Apply)**

**11.** **Outlier Detection:** Use the Z-score or IQR method to programmatically detect the "150-year-old user" in the dataset and filter them out. **(RBT Level 4: Analyze)**

---

### **Module 3: Grouping and Correlation**

**12.** Using the `groupby()` mechanics, find the top 3 directors based on their average user rating. Rename the resulting aggregation column to `Average_Score`. **(RBT Level 3: Apply)**

**13.** Create a **Pivot Table** that shows the average rating for each `genre` broken down by `gender`. What does this tell you about genre-specific audience preferences? **(RBT Level 4: Analyze)**

**14.** Define the difference between **Univariate**, **Bivariate**, and **Multivariate** analysis. Provide one example of each using the columns available in this movie dataset. **(RBT Level 2: Understand)**

**15.** Perform a bivariate analysis between `review_length` (calculated by word count) and `rating`. Calculate the correlation coefficient and explain if longer reviews tend to be more positive or negative. **(RBT Level 4: Analyze)**


---

## Solutions for EDA & Data Transformation Case Study

### **Module 1: Introduction to EDA**

**1. Measurement Scales Identification**

* **`movie_id`**: **Nominal** (Identifier only; math like averaging ID numbers is meaningless).
* **`genre`**: **Nominal** (Categories with no inherent mathematical order).
* **`rating`**: **Ordinal** (Ordered categories where 5 > 4, but the "distance" between them is subjective) or **Interval** if treated as a score.
* **`age`**: **Ratio** (Has a true zero point and mathematical operations are valid).


In [2]:
import pandas as pd
df = pd.read_csv('movie_ratings_dirty.csv')

print(df.head(10))

# Count missing values
null_counts = df.isnull().sum()
# Calculate percentage for rating
rating_null_pct = (df['rating'].isnull().sum() / len(df)) * 100

print(null_counts)
print(f"Missing Rating Percentage: {rating_null_pct:.2f}%")

  user_id  age      gender  movie_id                              title  \
0     U27  150        Male       101                          Inception   
1     U27   63        Male       104                 Mad Max: Fury Road   
2     U27   63        Male       112  Everything Everywhere All At Once   
3     U21   32  Non-binary       101                          Inception   
4     U21   32  Non-binary       103                           Parasite   
5     U21   32  Non-binary       107                      Spirited Away   
6     U21   32           M       115  Spider-Man: Into the Spider-Verse   
7     U20   31      Female       101                          Inception   
8     U20   31      Female       113                  Blade Runner 2049   
9     U29   62        Male       101                          Inception   

       genre           director  release_year  rating  \
0     Sci-Fi  Christopher Nolan          2010     3.0   
1     Action      George Miller          2015    10.0   
2  

This code snippet loads the `movie_ratings_dirty.csv` file into a pandas DataFrame named `df`. It then prints the first 10 rows of the DataFrame to give an initial view of the data structure and content. Following this, it calculates and displays the number of missing values (`NaN`) in each column using `df.isnull().sum()`. Finally, it specifically calculates the percentage of missing values in the `rating` column, which is crucial for understanding data completeness for a key analytical feature.

**3. Detecting Anomalies via Descriptive Statistics**

In [ ]:
# Identify outliers like age=150 or rating=10
print(df[['age', 'rating']].describe())

              age     rating
count   65.000000  59.000000
mean    44.507692   3.135593
std     21.899003   1.716629
min     19.000000   1.000000
25%     25.000000   2.000000
50%     50.000000   3.000000
75%     61.000000   4.500000
max    150.000000  10.000000


This code applies the `describe()` method to the `age` and `rating` columns. The output provides key descriptive statistics such as count, mean, standard deviation, minimum, maximum, and quartile values. This helps in identifying potential data anomalies, such as an `age` of 150 (which is unrealistic) or a `rating` of 10 (which might be outside an assumed scale of 1-5 or 1-7 depending on the dataset's context), indicating the need for data cleaning.

**4. Data Refactoring Significance**
Refactoring ensures that identifiers (`user_id`) are treated as objects/strings rather than integers. This prevents the model or statistical summary from calculating meaningless metrics like the "average user ID."



In [3]:
df['movie_id'] = df['movie_id'].astype(str)
df['user_id'] = df['user_id'].astype(str)
print(df[['movie_id', 'user_id']].dtypes)

movie_id    object
user_id     object
dtype: object


This code performs **data refactoring** by converting the `movie_id` and `user_id` columns to `object` (string) data types. This is important because these columns are unique identifiers and should not be treated as numerical data where statistical operations like calculating a mean would be nonsensical. By converting them to strings, we ensure they are handled appropriately for identification and categorical analysis, aligning with the principles of data refactoring for proper statistical analysis.

**5. Dropping Columns**

In [ ]:
# Drop director for demographic study
df_reduced = df.drop(columns=['director'])

This code demonstrates how to drop a specific column from the DataFrame using the `drop()` method. In this example, the `director` column is dropped, creating a new DataFrame `df_reduced`. This action is typically performed when a particular column is not relevant for a specific analysis (e.g., a user-centric demographic study where the director's information is not needed), thereby simplifying the dataset and reducing computational overhead.

---

### **Module 2: Data Transformation**

**6. Data Deduplication**

In [ ]:
print(f"Before: {len(df)}")
df = df.drop_duplicates()
print(f"After: {len(df)}")

Before: 65
After: 65


This cell performs data deduplication. It first prints the number of rows in the DataFrame before deduplication using `len(df)`. Then, `df.drop_duplicates()` is applied to remove any rows that are exact duplicates, updating the DataFrame `df`. Finally, it prints the number of rows after deduplication to show how many duplicates were removed. Deduplication is an essential step in data cleaning to ensure data integrity and prevent biased analysis due to redundant entries.

**7. Merging Dataframes**

To demonstrate how a combined DataFrame like our `df` is constructed from separate entities, let's simulate having distinct `df_users`, `df_movies`, and `df_reviews` DataFrames. We'll extract the relevant columns from our existing `df` to ensure consistency in data and column names.

In [9]:
# 1. Create a hypothetical df_users DataFrame
# Contains unique user information (user_id, age, gender)
df_users_reconstructed = df[['user_id', 'age', 'gender']].drop_duplicates().reset_index(drop=True)

# 2. Create a hypothetical df_movies DataFrame
# Contains unique movie information (movie_id, title, genre, director, release_year)
df_movies_reconstructed = df[['movie_id', 'title', 'genre', 'director', 'release_year']].drop_duplicates().reset_index(drop=True)

# 3. Create a hypothetical df_reviews DataFrame (the interaction table)
# Contains user-movie interactions and their specific rating/review (user_id, movie_id, rating, review_text)
df_reviews_reconstructed = df[['user_id', 'movie_id', 'rating', 'review_text']].copy()

print("Reconstructed df_users (sample):")
display(df_users_reconstructed.head())
print("\nReconstructed df_movies (sample):")
display(df_movies_reconstructed.head())
print("\nReconstructed df_reviews (sample):")
display(df_reviews_reconstructed.head())

Reconstructed df_users (sample):


,user_id,age,gender
0,U27,150,Male
1,U27,63,Male
2,U21,32,Non-binary
3,U21,32,M
4,U20,31,Female



Reconstructed df_movies (sample):


,movie_id,title,genre,director,release_year
0,101,Inception,Sci-Fi,Christopher Nolan,2010
1,104,Mad Max: Fury Road,Action,George Miller,2015
2,112,Everything Everywhere All At Once,Sci-Fi,Daniels,2022
3,103,Parasite,Drama,Bong Joon-ho,2019
4,107,Spirited Away,Animation,Hayao Miyazaki,2001



Reconstructed df_reviews (sample):


,user_id,movie_id,rating,review_text
0,U27,101,3.0,A bit confusing but very visually pleasing.
1,U27,104,10.0,The acting was superb but the plot was a bit m...
2,U27,112,2.0,Masterpiece. The cinematography was incredibel!
3,U21,101,2.0,"The pacing was a bit slow, but the ending made..."
4,U21,103,4.0,The acting was superb but the plot was a bit m...


Now that we have our separate DataFrames, we can demonstrate how they would be merged to form a structure similar to our original `df`. We'll typically start by merging the user information with the reviews, and then add the movie details.

In [10]:
# Step A: Merge df_users_reconstructed with df_reviews_reconstructed on 'user_id'
# This combines user demographic data with their reviews and ratings.
users_reviews_merged = pd.merge(
    df_users_reconstructed,
    df_reviews_reconstructed,
    on='user_id',
    how='inner' # Use inner join to keep only users who have left reviews
)

print("Users and Reviews Merged (sample):")
display(users_reviews_merged.head())

Users and Reviews Merged (sample):


,user_id,age,gender,movie_id,rating,review_text
0,U27,150,Male,101,3.0,A bit confusing but very visually pleasing.
1,U27,150,Male,104,10.0,The acting was superb but the plot was a bit m...
2,U27,150,Male,112,2.0,Masterpiece. The cinematography was incredibel!
3,U27,63,Male,101,3.0,A bit confusing but very visually pleasing.
4,U27,63,Male,104,10.0,The acting was superb but the plot was a bit m...


The previous step combined user demographics with their specific reviews and ratings. Now, we'll merge this result with the movie details to get a comprehensive view, similar to our original `df`.

In [11]:
# Step B: Merge the combined users_reviews_merged with df_movies_reconstructed on 'movie_id'
# This adds the movie's title, genre, director, and release year to each review entry.
final_reconstructed_df = pd.merge(
    users_reviews_merged,
    df_movies_reconstructed,
    on='movie_id',
    how='inner' # Use inner join to keep only reviews for existing movies
)

# Reorder columns to closely match the original df's structure for comparison
final_reconstructed_df = final_reconstructed_df[[
    'user_id', 'age', 'gender', 'movie_id', 'title', 'genre', 'director',
    'release_year', 'rating', 'review_text'
]]

print("Final Reconstructed DataFrame (sample, matching original df structure):")
display(final_reconstructed_df.head())

Final Reconstructed DataFrame (sample, matching original df structure):


,user_id,age,gender,movie_id,title,genre,director,release_year,rating,review_text
0,U27,150,Male,101,Inception,Sci-Fi,Christopher Nolan,2010,3.0,A bit confusing but very visually pleasing.
1,U27,150,Male,104,Mad Max: Fury Road,Action,George Miller,2015,10.0,The acting was superb but the plot was a bit m...
2,U27,150,Male,112,Everything Everywhere All At Once,Sci-Fi,Daniels,2022,2.0,Masterpiece. The cinematography was incredibel!
3,U27,63,Male,101,Inception,Sci-Fi,Christopher Nolan,2010,3.0,A bit confusing but very visually pleasing.
4,U27,63,Male,104,Mad Max: Fury Road,Action,George Miller,2015,10.0,The acting was superb but the plot was a bit m...


This `final_reconstructed_df` now closely mirrors the structure and content of our initial `df`, demonstrating how a single comprehensive DataFrame can be built by merging information from various logically separate tables representing users, movies, and their interactions.

In [ ]:
# Assuming you have separate df_users and df_movies
# df_merged = pd.merge(df_users, df_ratings, on='user_id', how='inner')

In [8]:
# Create sample df_users and df_movies DataFrames for demonstration with more attributes
df_users = pd.DataFrame({
    'user_id': ['U27', 'U21', 'U20', 'U29', 'U100'],
    'age': [63, 32, 31, 62, 25],
    'gender': ['Male', 'Non-binary', 'Female', 'Male', 'Female'],
    'city': ['New York', 'Los Angeles', 'Chicago', 'Houston', 'Miami'],
    'occupation': ['Engineer', 'Artist', 'Teacher', 'Retired', 'Student'],
    'membership_level': ['Gold', 'Silver', 'Gold', 'Bronze', 'Silver']
})

df_movies = pd.DataFrame({
    'movie_id': ['101', '104', '112', '103', '200'],
    'title': ['Inception', 'Mad Max: Fury Road', 'Everything Everywhere All At Once', 'Parasite', 'New Movie'],
    'director': ['Christopher Nolan', 'George Miller', 'Daniels', 'Bong Joon-ho', 'New Director'],
    'production_company': ['Warner Bros', 'Warner Bros', 'A24', 'CJ Entertainment', 'Indie Prod'],
    'critic_score': [8.8, 8.1, 8.5, 9.0, 7.5]
})

# Create a simplified df_ratings for demonstration, linking users and movies
df_ratings = df[['user_id', 'movie_id', 'rating']].copy()

print("Sample df_users (with more attributes):")
display(df_users)
print("\nSample df_movies (with more attributes):")
display(df_movies)
print("\nSample df_ratings (from original df):")
display(df_ratings.head())

Sample df_users (with more attributes):


,user_id,age,gender,city,occupation,membership_level
0,U27,63,Male,New York,Engineer,Gold
1,U21,32,Non-binary,Los Angeles,Artist,Silver
2,U20,31,Female,Chicago,Teacher,Gold
3,U29,62,Male,Houston,Retired,Bronze
4,U100,25,Female,Miami,Student,Silver



Sample df_movies (with more attributes):


,movie_id,title,director,production_company,critic_score
0,101,Inception,Christopher Nolan,Warner Bros,8.8
1,104,Mad Max: Fury Road,George Miller,Warner Bros,8.1
2,112,Everything Everywhere All At Once,Daniels,A24,8.5
3,103,Parasite,Bong Joon-ho,CJ Entertainment,9.0
4,200,New Movie,New Director,Indie Prod,7.5



Sample df_ratings (from original df):


,user_id,movie_id,rating
0,U27,101,3.0
1,U27,104,10.0
2,U27,112,2.0
3,U21,101,2.0
4,U21,103,4.0


First, let's demonstrate an **Inner Join**. An inner join returns only the rows that have matching values in both DataFrames being joined. In this case, it will combine user demographics with their ratings only for users and movies that exist in *both* `df_users` and `df_ratings`.

In [5]:
# Inner Join: Combine df_users with df_ratings based on user_id
inner_joined_df = pd.merge(df_users, df_ratings, on='user_id', how='inner')

print("Inner Joined DataFrame (df_users + df_ratings):")
display(inner_joined_df.head())

Inner Joined DataFrame (df_users + df_ratings):


,user_id,age,gender,city,movie_id,rating
0,U27,63,Male,New York,101,3.0
1,U27,63,Male,New York,104,10.0
2,U27,63,Male,New York,112,2.0
3,U21,32,Non-binary,Los Angeles,101,2.0
4,U21,32,Non-binary,Los Angeles,103,4.0


Next, let's demonstrate a **Left Join**. A left join returns all rows from the left DataFrame (`df_users` in this case) and the matching rows from the right DataFrame (`df_ratings`). If there's no match in the right DataFrame, it fills with `NaN` values for the columns from the right DataFrame. This is useful when you want to keep all user information, even if they haven't provided any ratings in `df_ratings`.

In [6]:
# Left Join: Combine df_users with df_ratings based on user_id
left_joined_df = pd.merge(df_users, df_ratings, on='user_id', how='left')

print("Left Joined DataFrame (df_users + df_ratings):")
display(left_joined_df.head())

Left Joined DataFrame (df_users + df_ratings):


,user_id,age,gender,city,movie_id,rating
0,U27,63,Male,New York,101,3.0
1,U27,63,Male,New York,104,10.0
2,U27,63,Male,New York,112,2.0
3,U21,32,Non-binary,Los Angeles,101,2.0
4,U21,32,Non-binary,Los Angeles,103,4.0


You can also join the `df_movies` with the `inner_joined_df` or `left_joined_df` to bring in movie details based on `movie_id`. Here's an example using the `left_joined_df` to include movie titles and directors for all rated movies, keeping all user and rating information:


In [7]:
# Further join with df_movies to add movie details
final_merged_df = pd.merge(left_joined_df, df_movies, on='movie_id', how='left')

print("Final Merged DataFrame (Users + Ratings + Movies using Left Join):")
display(final_merged_df.head())

Final Merged DataFrame (Users + Ratings + Movies using Left Join):


,user_id,age,gender,city,movie_id,rating,title,director
0,U27,63,Male,New York,101,3.0,Inception,Christopher Nolan
1,U27,63,Male,New York,104,10.0,Mad Max: Fury Road,George Miller
2,U27,63,Male,New York,112,2.0,Everything Everywhere All At Once,Daniels
3,U21,32,Non-binary,Los Angeles,101,2.0,Inception,Christopher Nolan
4,U21,32,Non-binary,Los Angeles,103,4.0,Parasite,Bong Joon-ho


This commented-out code snippet illustrates how one would merge two separate DataFrames (`df_users` and `df_ratings`) using pandas' `merge()` function. It shows an example of an `inner join` based on the common `user_id` column. Merging is crucial when data is distributed across multiple tables and needs to be combined for comprehensive analysis, such as linking user demographics with their movie ratings.

**8. Handling Missing Ratings**

In [ ]:
# (a) Drop
df_dropped = df.dropna(subset=['rating'])
# (b) Mean Imputation
df_mean = df.copy()
df_mean['rating'] = df_mean['rating'].fillna(df['rating'].mean())
# (c) Forward Fill
df_ffill = df['rating'].fillna(method='ffill')

/tmp/ipykernel_6362/2280679971.py:7: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df_ffill = df['rating'].fillna(method='ffill')


This code demonstrates three different techniques for handling missing values in the `rating` column:  
1.  **Dropping rows**: `df_dropped = df.dropna(subset=['rating'])` creates a new DataFrame where all rows with missing `rating` values are removed.  
2.  **Mean Imputation**: `df_mean['rating'] = df_mean['rating'].fillna(df['rating'].mean())` replaces missing `rating` values with the mean of the existing `rating` values.  
3.  **Forward Fill**: `df_ffill = df['rating'].fillna(method='ffill')` fills missing `rating` values with the last valid observation in the column. Each method has different implications for the dataset and its subsequent analysis.

**9. Categorical Standardization**

In [ ]:
df['genre'] = df['genre'].replace({'sci-fi': 'Sci-Fi'})
df['gender'] = df['gender'].replace({'M': 'Male'})

This code performs standardization on categorical columns to ensure consistency in data entry. It uses the `replace()` method to:  
-   Convert variations of 'sci-fi' to 'Sci-Fi' in the `genre` column.  
-   Convert 'M' to 'Male' in the `gender` column.  
This process is vital for accurate categorical analysis and visualization, as inconsistent spellings or abbreviations can lead to categories being treated as distinct when they should be unified.

**10. Discretization and Binning**

In [ ]:
bins = [0, 19, 35, 60, 100]
labels = ['Teen', 'Young Adult', 'Middle Aged', 'Senior']
df['age_group'] = pd.cut(df['age'], bins=bins, labels=labels)

This code demonstrates **discretization** by creating a new categorical column `age_group` from the numerical `age` column. It defines `bins` (age ranges) and corresponding `labels` ('Teen', 'Young Adult', 'Middle Aged', 'Senior'). The `pd.cut()` function then assigns each `age` value to its appropriate age group. This technique transforms continuous numerical data into discrete categories, which can be useful for analyses requiring grouped age data or for simplifying complex numerical distributions.

**11. Outlier Filtering (IQR Method)**

In [ ]:
Q1 = df['age'].quantile(0.25)
Q3 = df['age'].quantile(0.75)
IQR = Q3 - Q1
df_clean = df[~((df['age'] < (Q1 - 1.5 * IQR)) | (df['age'] > (Q3 + 1.5 * IQR)))]

**Addressing Inconsistencies for User Profiles**

Based on your observations, we need to correct specific inconsistencies in the `df` before creating user-centric DataFrames like `df_users_reconstructed`. This involves:
1.  **Correcting Anomalous Age:** `user_id` U27 has an age of 150, which is an outlier. We will standardize this to 63, which is also present for this user and more realistic.
2.  **Standardizing Gender:** `user_id` U21 appears with both 'Non-binary' and 'Male' genders. For consistency in a user profile, we will standardize all entries for U21 to 'Non-binary'.

In [12]:
# 1. Correct Age for user_id 'U27': Replace 150 with 63
df.loc[(df['user_id'] == 'U27') & (df['age'] == 150), 'age'] = 63

# 2. Standardize Gender for user_id 'U21': Set to 'Non-binary'
df.loc[df['user_id'] == 'U21', 'gender'] = 'Non-binary'

print("DataFrame 'df' updated for U27 age and U21 gender inconsistencies.")

DataFrame 'df' updated for U27 age and U21 gender inconsistencies.


Now that the `df` has been corrected for these specific user profile inconsistencies, we will re-generate `df_users_reconstructed` to reflect these changes. The `df_movies_reconstructed` and `df_reviews_reconstructed` do not need to be re-generated as the corrections were user-specific.

In [13]:
# Re-create df_users_reconstructed with corrected df
df_users_reconstructed = df[['user_id', 'age', 'gender']].drop_duplicates().reset_index(drop=True)

print("Reconstructed df_users (sample) after corrections:")
display(df_users_reconstructed.head())

Reconstructed df_users (sample) after corrections:


,user_id,age,gender
0,U27,63,Male
1,U21,32,Non-binary
2,U20,31,Female
3,U29,62,Male
4,U28,59,Female


This code demonstrates **outlier detection and filtering** using the Interquartile Range (IQR) method for the `age` column.  
-   It calculates the first quartile (Q1) and third quartile (Q3) of the `age` data.  
-   The IQR is then computed as the difference between Q3 and Q1.  
-   Outliers are identified as values falling below `Q1 - 1.5 * IQR` or above `Q3 + 1.5 * IQR`.  
-   Finally, `df_clean` is created, containing only the rows where `age` is within the calculated acceptable range, effectively removing outliers like the '150-year-old user'.

---

### **Module 3: Grouping and Correlation**

**12. Groupby Aggregation**

In [ ]:
director_ranking = df.groupby('director')['rating'].mean().rename("Average_Score").sort_values(ascending=False)
print(director_ranking.head(3))

director
Denis Villeneuve    4.20
Martin McDonagh     4.00
George Miller       3.75
Name: Average_Score, dtype: float64


This code uses the `groupby()` method to calculate the average `rating` for each `director`.  
-   `df.groupby('director')['rating'].mean()` groups the DataFrame by director and then calculates the mean rating for each group.  
-   `.rename("Average_Score")` renames the resulting series to `Average_Score`.  
-   `.sort_values(ascending=False)` sorts the directors by their average score in descending order.  
-   `print(director_ranking.head(3))` displays the top 3 directors based on this average rating. This helps in identifying the most critically acclaimed directors according to user ratings.

**13. Pivot Table Analysis**

In [ ]:
pivot = df.pivot_table(index='genre', columns='gender', values='rating', aggfunc='mean')
print(pivot)

gender       Female      Male  Non-binary
genre                                    
Action     2.000000  7.500000        4.00
Animation  2.000000  1.000000         NaN
Crime      3.000000  1.500000        5.00
Drama      3.571429  4.000000        3.50
Horror          NaN  4.000000        1.00
Sci-Fi     3.375000  3.857143        1.75


This code creates a **pivot table** to analyze the average `rating` across different `genre` and `gender` combinations.  
-   `df.pivot_table()` is used with `index='genre'` (rows), `columns='gender'` (columns), `values='rating'` (data to aggregate), and `aggfunc='mean'` (aggregation function).  
-   The resulting pivot table `pivot` shows the average rating for each genre, broken down by gender, providing insights into genre-specific audience preferences across different gender groups.

**14. Types of Analysis Examples**

To better understand the differences, let's look at examples using our movie dataset:

*   **Univariate Analysis:** This involves analyzing a single variable. For instance, creating a histogram to visualize the distribution of `ratings` to understand how frequently each rating appears. This helps in grasping the central tendency and spread of individual features.
    *   _Example:_ Histogram of `ratings`.

*   **Bivariate Analysis:** This explores the relationship between two variables. A scatter plot of `age` vs. `rating` can show if there's any pattern or correlation between a user's age and the ratings they give. This helps in understanding how two variables move together.
    *   _Example:_ Scatter plot of `age` vs `rating`.

*   **Multivariate Analysis:** This examines the relationships among three or more variables simultaneously. A heatmap showing the correlation between `age`, `release_year`, and `rating` could reveal complex interactions, such as how both age and release year might influence ratings. This is useful for uncovering more intricate patterns within the data.
    *   _Example:_ Heatmap showing correlation between `age`, `release_year`, and `rating`.

In [ ]:
# Feature engineering
df['review_length'] = df['review_text'].str.split().str.len()

# Correlation (ignoring NaNs)
correlation = df['review_length'].corr(df['rating'])
print(f"Correlation: {correlation}")

Correlation: 0.2678157688184501


This code performs a **bivariate analysis** to understand the relationship between review length and movie rating.  
-   **Feature Engineering**: A new column `review_length` is created by calculating the word count of each `review_text` using `df['review_text'].str.split().str.len()`.  
-   **Correlation Calculation**: `df['review_length'].corr(df['rating'])` calculates the Pearson correlation coefficient between `review_length` and `rating`, ignoring any NaN values.  
-   The output `Correlation: 0.2678157688184501` suggests a weak positive correlation, indicating that slightly longer reviews tend to be associated with slightly higher ratings, although the relationship is not strong.